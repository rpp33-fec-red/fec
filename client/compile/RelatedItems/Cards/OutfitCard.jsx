import React from 'react';
import '../relatedItems.scss';

class OutfitCard extends React.Component {
    constructor (props) {
        super (props);
        this.state = {
            id: props.outfit_id,
            photo: `https://i.pinimg.com/originals/2d/38/62/2d3862abf214ecdb6a1c231e035c1796.jpg`,
            category: 'Pug',
            product_name: '',
            default_price: '$$$',
        }
        this.getOutfitInfo = this.getOutfitInfo.bind(this);
    }
    componentDidMount () {
        this.getOutfitInfo(this.props.outfit_id);
    }

    getOutfitInfo (id) {

        this.props.getData(['products',id], (data) => {
            this.setState ({
                category: data.results.category,
                product_name: data.results.name,
                default_price: data.results.default_price
            });
        });

        this.props.getData(['products',id,'styles'],(data) => {
            this.setState({
                photo: data.results.results[0]['photos'][0]['thumbnail_url']
            })
        });
    }

    render () {
        return (
            <div className="productCard" >
                <button
                    id="deleteButton"
                    value={this.state.id}
                    onClick={this.props.handleDelete}
                    >X</button>
                <img src={this.state.photo} />
                <p className="category" ></p>
                <a
                    className={`cardtitle ${this.props.product_id}`}
                    onClick={this.props.handleClick}
                    href={`/products/${this.props.product_id}`}
                ><b>{this.state.product_name}</b></a>
                <p className="default-price" >$ {this.state.default_price}</p>

            </div>
        );
    }
}

export default OutfitCard;
