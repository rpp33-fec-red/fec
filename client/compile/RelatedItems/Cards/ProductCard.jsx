/* eslint-disable react/react-in-jsx-scope */
import StarsComponent from '../../showStars';
import '../relatedItems.scss';
import PropTypes from 'prop-types';

// eslint-disable-next-line no-undef
class ProductCard extends React.Component {
  constructor (props) {
    super (props);
    this.state = {
      photo: `https://i.pinimg.com/originals/2d/38/62/2d3862abf214ecdb6a1c231e035c1796.jpg`,
      category: 'Pug',
      product_name: '',
      default_price: '$$$',
      sale_price: null,
      original_price: '$$'
    };
    this.getCardInfo = this.getCardInfo.bind(this);
  }

  componentDidMount () {
    this.getCardInfo(this.props.product_id);
  }

  getCardInfo (id) {

    this.props.getData(['products', id, ''], (data) => {
      this.setState ({
        category: data.results.category,
        product_name: data.results.name,
        default_price: data.results.default_price
      });
    });

    this.props.getData(['products', id, 'styles'], (data) => {
      this.setState({
        photo: data.results.results[0]['photos'][0]['thumbnail_url'],
        original_price: data.results.results[0]['original_price'],
        sale_price: parseInt(data.results.results[0]['sale_price'])
      });
    });
  }

  render () {
    return (
      <div className={`productCard ${this.props.product_id}`}>
        <button
          id = "compareButton"
          value = {this.props.product_id}
          onClick = {this.props.handleCompare}
        >*</button>
        <a href={`/?productid=${this.props.product_id}`} style={{"text-decoration":"none", "color":"black"}}>
          <img src={this.state.photo} alt="Photo" className="productImage" />
          <p className="product_category" >{this.state.category}</p>
          <p className='product_cardtitle'>{this.state.product_name}</p>
          <p className="product_default_price">
            { !this.state.sale_price
              ? (`$ ${this.state.default_price}`)
              : (<span><span className='product_original_price'>$ {this.state.original_price}</span><b><span className='product_sale_price'>  $ {this.state.sale_price}</span></b></span>)
            }</p>
          <StarsComponent product_id={this.props.product_id} />
        </a>
      </div>
    );
  }
}

ProductCard.propTypes = {
  getData: PropTypes.any,
  product_id: PropTypes.any,
  handleClick: PropTypes.any,
  handleCompare: PropTypes.any
};
  
export default ProductCard;