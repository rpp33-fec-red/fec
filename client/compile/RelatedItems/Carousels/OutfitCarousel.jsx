/* eslint-disable react/react-in-jsx-scope */
import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import OutfitCard from '../Cards/OutfitCard.jsx';
import '../relatedItems.scss';

const useWindowSize = () => {
  const isWindow = typeof window === 'object';
  const [windowSize, setWindowSize] = useState (isWindow ? window.innerWidth : undefined);
  useEffect(() => {
    const setSize = () => {
      setWindowSize(window.innerWidth);
    };
    if (isWindow) {
      window.addEventListener("resize", setSize);
      return () => window.removeEventListener("resize", setSize);
    }
  }, [isWindow, setWindowSize]);
  return windowSize;
};

const chunkArray = (array, chunkSize) => {
  let sets = [];
  for (let i = 0; i < array.length; i+=chunkSize) {
    let chunk = array.slice(i, i+chunkSize);
    sets.push(chunk);
  }
  return sets;
};

const OutfitCarousel = (props) => {
  let outfit_Ids = props.outfit_Ids;
  let sets = chunkArray(outfit_Ids, Math.floor(useWindowSize()/400));
  let numberOfSets = sets.length;
  let [current, setCurrent] = useState(0);
  const handleLeft = () => {return current === 0 ? setCurrent(numberOfSets - 1) : setCurrent(current - 1);};
  const handleRight = () => { return current === numberOfSets - 1 ? setCurrent(0) : setCurrent(current + 1);};

  return (
    <div className="outfitCarouselContainer">
      <div className="productCarousel">
        <div className="productCard" >
          <button
            id="addButton"
            varian="outline-primary"
            onClick={props.handleAddToOutfit}
          > + </button>
        </div>

        {current === 0 
          ? (<></>)
          : (numberOfSets > 0 && <button className="chevron-left" onClick={handleLeft}><svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" fill="currentColor" className="bi bi-chevron-left" viewBox="0 0 16 16"><path fillRule="evenodd" d="M11.354 1.646a.5.5 0 0 1 0 .708L5.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z"/></svg></button>)
        }
        {sets[current] && sets[current].map((id) => {
          return (
            <OutfitCard
              key={id}
              index={id}
              outfit_id={id}
              handleDelete={props.handleDelete}
              getData={props.getData}
            />
          );
        })}
        {current === 0 
          ? (numberOfSets > 0 && <button className="chevron-right" onClick={handleRight}><svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" fill="currentColor" className="bi bi-chevron-right" viewBox="0 0 16 16"><path fillRule="evenodd" d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"/></svg></button>)
          : (<></>)
        }
      </div>
    </div>
  );
};
OutfitCarousel.propTypes = {
  outfit_Ids: PropTypes.any,
  getData: PropTypes.any,
  outfitLoaded: PropTypes.any,
  handleAddToOutfit: PropTypes.any,
  handleDelete: PropTypes.any
};
export default OutfitCarousel;
