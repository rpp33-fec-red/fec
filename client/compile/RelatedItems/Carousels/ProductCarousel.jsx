import React, {useState, useEffect} from 'react';
import ProductCard from '../Cards/ProductCard.jsx';
import PropTypes from 'prop-types';
import '../relatedItems.scss';

const useWindowSize = () => {
  const isWindow = typeof window === 'object';
  const [windowSize, setWindowSize] = useState (isWindow ? window.innerWidth : undefined);
  useEffect(() => {
    const setSize = () => {
      setWindowSize(window.innerWidth);
    };
    if (isWindow) {
      window.addEventListener("resize", setSize);
      return () => window.removeEventListener("resize", setSize);
    }
  }, [isWindow, setWindowSize]);
  return windowSize;
};

const chunkArray = (array, chunkSize) => {
  let sets = [];
  for (let i = 0; i < array.length; i+=chunkSize) {
    let chunk = array.slice(i, i+chunkSize);
    sets.push(chunk);
  }
  return sets;
};

const ProductCarousel = (props) => {
  let relatedProducts = props.relatedProducts;
  let sets = chunkArray(relatedProducts, Math.floor(useWindowSize()/300));
  console.log('window size', useWindowSize());
  let numberOfSets = sets.length;
  let [current, setCurrent] = useState(0);
  const handleLeft = () => {return current === 0 ? setCurrent(numberOfSets - 1) : setCurrent(current - 1);};
  const handleRight = () => { return current === numberOfSets - 1 ? setCurrent(0) : setCurrent(current + 1);};
  return (
    <div className="relatedProductCarouselContainer">
      <div className="productCarousel">
        {current === 0 
          ? (<></>)
          : (numberOfSets > 0 && <button className="chevron-left" onClick={handleLeft}><svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" fill="currentColor" className="bi bi-chevron-left" viewBox="0 0 16 16"><path fillRule="evenodd" d="M11.354 1.646a.5.5 0 0 1 0 .708L5.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z"/></svg></button>)
        }
        {sets[current] && sets[current].map((id) => {
          return (
            <ProductCard 
              key={id}
              product_id={id}
              getData={props.getData}
              handleClick={props.handleClick}
              handleCompare={props.handleCompare}
            /> 
          );
        })}
        {current === 0 
          ? (numberOfSets > 0 && <button className="chevron-right" onClick={handleRight}><svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" fill="currentColor" className="bi bi-chevron-right" viewBox="0 0 16 16"><path fillRule="evenodd" d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"/></svg></button>)
          : (<></>)
        }
      </div>
    </div>
  );
};

ProductCarousel.propTypes = {
  getData: PropTypes.any,
  relatedProducts: PropTypes.any,
  handleClick: PropTypes.any,
  handleCompare: PropTypes.any
};
  
export default ProductCarousel;