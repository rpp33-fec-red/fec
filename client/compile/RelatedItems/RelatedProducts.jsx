import React from 'react';
import ComparisonModal from './ComparisonModal.jsx';
import ProductCarousel from './Carousels/ProductCarousel.jsx';
import OutfitCarousel from './Carousels/OutfitCarousel.jsx';
import PropTypes from 'prop-types';

class RelatedProducts extends React.Component {
  constructor (props) {
    super (props);
    this.state = {
      current: {},
      compare: {},
      relatedProducts: [], 
      outfitIds: [],
      outfitLoaded: false,
      showModal: false
    };
    this.removeDuplicate = this.removeDuplicate.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.getOutfits = this.getOutfits.bind(this);
    this.handleAddToOutfit = this.handleAddToOutfit.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleCompare = this.handleCompare.bind(this);
    this.toggleState = this.toggleState.bind(this);
  }

  //all 64621 needs to be change to props.productId once params.id is passed to component
  componentDidMount () {
    this.props.getData(['products', this.props.product_id, 'related'], (data) => {
      const uniqueIds = this.removeDuplicate(data.results);
      this.setState({ relatedProducts: uniqueIds });
    });

    this.props.getData(['products',this.props.product_id, ''], (data) => {
      this.setState ({ current: data.results });
    });

    this.getOutfits();
  }

  removeDuplicate (array) {
    // eslint-disable-next-line no-undef
    return [...new Set(array)];
  }

  handleClick (e) {
    let id = e.currentTarget.className.split(' ')[2];
    this.props.getData (['products', id, 'related'], (data) => {
      const uniqueIds = this.removeDuplicate(data.results);
      this.setState({ relatedProducts: uniqueIds });
    });
    this.props.getData(['products',id, ''], (data) => {
      this.setState ({ current: data.results });
    });
  }

  handleCompare (e) {
    let id = e.target.value;
    this.props.getData (['products', id, ''], (data) => {
      this.setState({ compare: data.results });
      this.setState({ showModal: true });
    });
  }

  toggleState () {
    this.setState ({ showModal: !this.state.showModal });
  }

  getOutfits () {
    let outfitIds = JSON.parse(localStorage.getItem("outfit"));
    if (outfitIds) {
      this.setState({ outfitIds: outfitIds});
      this.setState({ outfitLoaded: true });
    }
  }

  handleAddToOutfit () {
    let currentOutfit = this.state.outfitIds.slice();
    console.log('currentOutfit', currentOutfit);
    if (currentOutfit.indexOf(this.state.current.id) === -1 && this.state.current.id !== null) {
      this.setState({ outfitLoaded: false }, () => {
        currentOutfit.unshift(this.state.current.id);
        this.setState({ outfitIds: currentOutfit }, () => {
          localStorage.setItem("outfit", JSON.stringify(currentOutfit));
          this.getOutfits();
        });
      });

    }
  }

  handleDelete (e) {
    let id = Number(e.target.value);
    let storage = this.state.outfitIds.slice();
    let index = storage.indexOf(id);
    storage.splice(index, 1);
    localStorage.setItem("outfit", JSON.stringify(storage));
    this.getOutfits();
  }

  render () {
    return (
      <div className='relatedProducts'>
        <br></br>
        <h2>Related Products</h2>
        {this.state.showModal
          ? (<ComparisonModal 
            current={this.state.current}
            compare={this.state.compare}
            show={this.state.showModal}
            onClose={ this.toggleState }
          />)
          : (<div></div>)
        }
        <ProductCarousel
          relatedProducts={this.state.relatedProducts}
          getData={this.props.getData}
          handleClick={this.handleClick}
          handleCompare={this.handleCompare}
        />
        <br></br>
        <h2>Your Outfit</h2>
        <OutfitCarousel
          outfit_Ids={this.state.outfitIds}
          getData={this.props.getData}
          outfitLoaded={this.state.outfitLoaded}
          handleAddToOutfit={this.handleAddToOutfit}
          handleDelete={this.handleDelete}
        />
        <br></br>

      </div>
    );
  }
}


RelatedProducts.propTypes = {
  getData: PropTypes.any,
  product_id: PropTypes.any
};

export default RelatedProducts;