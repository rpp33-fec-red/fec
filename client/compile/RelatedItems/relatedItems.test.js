/* eslint-disable no-undef */


import React from 'react';
import { render, screen, fireEvent, getByRole } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-environment-jsdom';
import TestRenderer from 'react-test-renderer';
import RelatedProducts from './RelatedProducts.jsx';
import ProductCarousel from './Carousels/ProductCarousel.jsx';
import OutfitCarousel from './Carousels/OutfitCarousel.jsx';
import ProductCard from './Cards/ProductCard.jsx';
import OutfitCard from './Cards/OutfitCard.jsx';

describe ('Snapshot of whole widget', () => {
  const testRenderer = TestRenderer.create(<RelatedProducts getData={()=>{}} product_id='64621'/> );
  expect (testRenderer.toJSON()).toMatchSnapshot();
});

describe ('Testing initial render', () => {
  test ('Related Product render', () => {
    render (<RelatedProducts getData={()=>{}} product_id='64621'/> );
    screen.debug();
  });

  test ('Product Carousel render', () => {
    render (<ProductCarousel
      relatedProducts={[]}
      getData={()=>{}}
      handleClick={()=>{}}
      handleCompare={()=>{}}
    /> );
    screen.debug();
  });

  test ('Outfit Carousel render', () => {
    render (<OutfitCarousel
      outfit_Ids={[]}
      getData={()=>{}}
      outfitLoaded={{}}
      handleAddToOutfit={()=>{}}
      handleDelete={()=>{}}
    /> );
    screen.debug();
  });

  test ('Product Card initial render', () => {
    render (<ProductCard 
      key='64621'
      product_id='64621'
      getData={()=>{}}
      handleClick={()=>{}}
      handleCompare={()=>{}}
    /> );
    screen.debug();
  });

  test ('Outfit Card initial render', () => {
    render (<OutfitCard
      key='64621'
      index='64621'
      outfit_id='64621'
      handleDelete={()=>{}}
      getData={()=>{}}
    /> );
    screen.debug();
  });
});

describe ('Testing clicks', () => {
  test ('should have href link with product', async () => {
    render (<ProductCard 
      key='64621'
      product_id='64621'
      getData={()=>{}}
      handleClick={()=>{}}
      handleCompare={()=>{}}
    />);
    let link = screen.getByRole('link');
    expect(link).toHaveAttribute("href", "/?productid=64621");
  });

  test ('* button on Product Card', async () => {
    let buttonText;
    render (<ProductCard 
      key='64621'
      product_id='64621'
      getData={()=>{}}
      handleClick={()=>{}}
      handleCompare={()=>{ buttonText='Clicked'; }}
    />);
    buttonText = screen.getByText('*');
    await fireEvent.click(screen.getByRole('button'));
    expect(buttonText).toBe('Clicked');
  });

  test ('X button on Outfit Card', async () => {
    let buttonText;
    render (<OutfitCard
      key='64621'
      index='64621'
      outfit_id='64621'
      handleDelete={()=>{ buttonText='Clicked'; }}
      getData={()=>{}}
    />);
    buttonText = screen.getByText('X');
    await fireEvent.click(screen.getByRole('button'));
    expect(buttonText).toBe('Clicked');
  });

});




