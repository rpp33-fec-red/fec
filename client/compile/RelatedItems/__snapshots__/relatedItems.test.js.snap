// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Products Component renders initially 1`] = `
<div
  className="relatedProducts"
>
  <br />
  <h2>
    Related Products
  </h2>
  <div />
  <div
    className="relatedProductCarouselContainer"
  >
    <div
      className="productCarousel"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "left",
            "padding": "2rem",
            "width": 0,
          }
        }
      />
    </div>
  </div>
  <br />
  <h2>
    Your Outfit
  </h2>
  <div
    className="outfitCarouselContainer"
  >
    <div
      className="productCarousel"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "left",
            "width": 400,
          }
        }
      >
        <div
          className="productCard"
        >
          <button
            id="addButton"
            onClick={[Function]}
            varian="outline-primary"
          >
             + 
          </button>
        </div>
        <div
          className="productCard"
        >
          <button
            id="deleteButton"
            onClick={[Function]}
            value={64622}
          >
            X
          </button>
          <img
            alt="Photo"
            className="productImage"
            src="https://i.pinimg.com/originals/2d/38/62/2d3862abf214ecdb6a1c231e035c1796.jpg"
          />
          <p
            className="category"
          />
          <a
            className="cardtitle 64622"
          >
            <b>
              
            </b>
          </a>
          <p
            className="default-price"
          >
            $ 
            $$$
          </p>
        </div>
      </div>
    </div>
  </div>
  <br />
</div>
`;
